csharp_dotnetcore/19.Bot-Authentication/AuthenticationBot.cs

```diff
@@ -0,0 +1,249 @@
+    /// API and the Outlook API to demonstrate making calls to a service that requires
+    /// authentication. Bot developers no longer need to host OAuth controllers or
+    /// manage the token life-cycle, as all of this can now be done by the Azure Bot Service.
+    /// </summary>
+    public class AuthenticationBot : IBot
+    {
+        // Your connection name
+        private const string ConnectionSettingName = "AADv2Connection";
+
+        // Instructions for the user with information about commands that this bot may handle.
+        private const string HelpText =
+            "You can type 'send <recipient_email>' to send an email, 'recent' to view recent unread mail" +
+            " 'me' to see information about yourself, or 'help' to view the commands" +
+            " again. Any other text will display your token.";
+
+        private readonly AuthenticationBotAccessors _stateAccessors;
+        private readonly DialogSet _dialogs;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AuthenticationBot"/> class.
+        /// In the constructor for the bot we are instantiating our <see cref="DialogSet"/>, giving our field a value,
+        /// adding our <see cref="WaterfallDialog"/> and <see cref="ChoicePrompt"/> to the dialog set.
+        /// We are also adding  multiple <see cref="WaterfallStep"/> to our <see cref="WaterfallDialog"/>.
+        /// </summary>
+        /// <param name="accessors">State accessors for the bot.</param>
+        public AuthenticationBot(AuthenticationBotAccessors accessors)
+        {
+            this._stateAccessors = accessors;
+            this._dialogs = new DialogSet(this._stateAccessors.ConversationDialogState);
+            this._dialogs.Add(OAuthHelpers.Prompt(ConnectionSettingName));
+            this._dialogs.Add(new ChoicePrompt("choicePrompt"));
+            this._dialogs.Add(new WaterfallDialog("graphDialog", new WaterfallStep[] { PromptStepAsync, ProcessStepAsync }));
+        }
+
+        /// <summary>
+        /// This controls what happens when an <see cref="Activity"/> gets sent to the bot.
+        /// </summary>
+        /// <param name="turnContext">Provides the <see cref="ITurnContext"/> for the turn of the bot.</param>
+        /// <param name="cancellationToken" >(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> representing the operation result of the operation.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            DialogContext dc = null;
+
+            switch (turnContext.Activity.Type)
+            {
+                case ActivityTypes.Message:
+
+                    dc = await ProcessInputAsync(turnContext, dc, cancellationToken);
```

Comment:
Why are you passing in `dc` which is a `null` value at this point? [`ProcessInputAsync()` immediately creates a DialogContext and assigns the value to `dc`](https://github.com/Microsoft/BotBuilder-Samples/pull/332/files#diff-b99c9232b9023815898ad8b91e46aa02R143), meaning that even if `dc` were not a null value, its actual value would be irrelevant.